### PR TITLE
Hallway & LSTM Improvements

### PR DESCRIPTION
**Pull request by awjuliani**
* Add `Hallway` example environment to validate LSTM models.
* Properly mask gradient for padded experiences when using `sequence_length > 1`
* Add support for feeding previous action into LSTM models to enable partial meta-learning.

### CODE REVIEW

File: docs/Learning-Environment-Examples.md [hunk 1]
```diff
@@ -57,10 +59,10 @@ If you would like to contribute environments, please see our
     * -0.01 for every step.
     * +1.0 if the agent navigates to the goal position of the grid (episode ends).
     * -1.0 if the agent navigates to an obstacle (episode ends).
-* Brains: One brain with the following state/action space.
-    * State space: None
+* Brains: One brain with the following observation/action space.
+    * Observation space: None
```
Comment (vincentpierre) [inline docs/Learning-Environment-Examples.md]: Vector Observation space

File: python/trainer_config.yaml [hunk 2]
```diff
@@ -1,29 +1,114 @@
...
+    time_horizon: 32
+    batch_size: 32
+    buffer_size: 320
+
+Ball3DBrainR:
     summary_freq: 1000
     normalize: true
-    batch_size: 1000
-    buffer_size: 10000 
+    # batch_size: 1000
```
Comment (eshvk) [inline python/trainer_config.yaml]: delete commented lines

File: python/unitytrainers/models.py [hunk 3]
```diff
@@ -29,6 +29,10 @@ def create_global_steps():
         increment_step = tf.assign(global_step, tf.add(global_step, 1))
         return global_step, increment_step
 
+    @staticmethod
+    def swish(input_activation):
+        return tf.multiply(input_activation, tf.nn.sigmoid(input_activation))
```
Comment (vincentpierre) [inline python/unitytrainers/models.py]: Can we have an inline comment for this?

File: python/unitytrainers/models.py [hunk 4]
```diff
@@ -112,6 +116,10 @@ def create_discrete_state_encoder(self, s_size, h_size, activation, num_layers):
     def create_new_obs(self, num_streams, h_size, num_layers, activation_fn):
         brain = self.brain
         s_size = brain.state_space_size * brain.stacked_states
+        if brain.action_space_type == "continuous":
+            activation_fn = tf.nn.tanh
+        else:
+            activation_fn = tf.nn.tanh
```
Comment (awjuliani) [inline python/unitytrainers/models.py]: This should be swish!

File: python/unitytrainers/models.py [hunk 5]
```diff
@@ -229,10 +240,10 @@ def create_cc_actor_critic(self, h_size, num_layers):
         a = tf.exp(-1 * tf.pow(tf.stop_gradient(self.output) - self.mu, 2) / (2 * self.sigma_sq))
         b = 1 / tf.sqrt(2 * self.sigma_sq * np.pi)
         self.all_probs = tf.multiply(a, b, name="action_probs")
-        self.probs = tf.identity(self.all_probs)
+        self.probs = tf.reduce_sum(self.all_probs, axis=1)
         self.entropy = tf.reduce_sum(0.5 * tf.log(2 * np.pi * np.e * self.sigma_sq))
```
Comment (vincentpierre) [inline python/unitytrainers/models.py]: I want tests on that. This does not seem right

File: python/unitytrainers/ppo/models.py [hunk 6]
```diff
@@ -57,17 +58,28 @@ def create_ppo_optimizer(self, probs, old_probs, value, entropy, beta, epsilon,
...
-        r_theta = probs / (old_probs + 1e-10)
-        p_opt_a = r_theta * self.advantage
-        p_opt_b = tf.clip_by_value(r_theta, 1 - decay_epsilon, 1 + decay_epsilon) * self.advantage
-        self.policy_loss = -tf.reduce_mean(tf.minimum(p_opt_a, p_opt_b))
-        self.value_loss = tf.reduce_mean(tf.squared_difference(self.returns_holder, tf.reduce_sum(value, axis=1)))
-        self.loss = self.policy_loss + 0.5 * self.value_loss - decay_beta * tf.reduce_mean(entropy)
+        self.mask = tf.equal(self.mask_input, 1.0)
+
+        vpredclipped = self.old_value + tf.clip_by_value(tf.reduce_sum(value, axis=1) - self.old_value,
+                                                         - decay_epsilon, decay_epsilon)
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/models.py]: Can we think of a better name for this variable ?

File: unity-environment/Assets/ML-Agents/Examples/Hallway/Scripts/HallwayAcademy.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    public Material failMaterial; //when fail, the ground will use this material for a few seconds. 
+	public float gravityMultiplier; //use ~3 to make things less floaty
+
+	public override void AcademyReset()
+	{
+
+    }
+	// Use this for initialization
+	void Start () {
+		Physics.gravity *= gravityMultiplier; //Normally you shouldn't override Start() or Awake() with MLAgents, but Start() isn't used in Academy.cs so this should be ok for now.
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Hallway/Scripts/HallwayAcademy.cs]: Can you use Initialize Academy instead ?

File: unity-environment/Assets/ML-Agents/Examples/Hallway/Scripts/HallwayAcademy.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+	}
+	
+	// Update is called once per frame
+	void Update () {
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Hallway/Scripts/HallwayAcademy.cs]: Can you remove this since it is not implemented

File: unity-environment/Assets/ML-Agents/Examples/Hallway/Scripts/HallwayAgent.cs [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    Rigidbody shortBlockRB;  //cached on initialization
+    Rigidbody agentRB;  //cached on initialization
+    Material groundMaterial; //cached on Awake()
+    Renderer groundRenderer;
+    HallwayAcademy academy;
+
+    int selection;
+
+    void Awake()
+    {
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Hallway/Scripts/HallwayAgent.cs]: Please use InitializeAgent.

File: python/trainer_config.yaml [hunk 10]
```diff
@@ -2,14 +2,14 @@ default:
...
-    buffer_size: 10000
+    buffer_size: 10240
     epsilon: 0.2
     gamma: 0.99
     hidden_units: 128
     lambd: 0.95
     learning_rate: 3.0e-4
     max_steps: 5.0e4
-    normalize: false
+    normalize: true
```
Comment (vincentpierre) [inline python/trainer_config.yaml]: Does this help in all cases ? Should the default really be true ?

Comment (awjuliani) [reply python/trainer_config.yaml]: You are right. Better to set it to false by default.

File: docs/Learning-Environment-Examples.md [hunk 11]
```diff
@@ -39,11 +41,11 @@ If you would like to contribute environments, please see our
 * Agent Reward Function: 
     * +0.1 for every step the ball remains on the platform. 
     * -1.0 if the ball falls from the platform.
-* Brains: One brain with the following state/action space.
-    * State space: (Continuous) 8 variables corresponding to rotation of platform, and position, rotation, and velocity of ball.
-    * State space (Hard Version): (Continuous) 5 variables corresponding to rotation of platform and position and rotation of ball.
+* Brains: One brain with the following observation/action space.
+    * Vector Observation space: (Continuous) 8 variables corresponding to rotation of platform, and position, rotation, and velocity of ball.
+    * Observation space (Hard Version): (Continuous) 5 variables corresponding to rotation of platform and position and rotation of ball.
```
Comment (vincentpierre) [inline docs/Learning-Environment-Examples.md]: Vector Observation

File: python/unitytrainers/models.py [hunk 12]
```diff
@@ -229,10 +242,10 @@ def create_cc_actor_critic(self, h_size, num_layers):
         a = tf.exp(-1 * tf.pow(tf.stop_gradient(self.output) - self.mu, 2) / (2 * self.sigma_sq))
         b = 1 / tf.sqrt(2 * self.sigma_sq * np.pi)
         self.all_probs = tf.multiply(a, b, name="action_probs")
-        self.probs = tf.identity(self.all_probs)
+        self.probs = tf.reduce_prod(self.all_probs, axis=1)
```
Comment (vincentpierre) [inline python/unitytrainers/models.py]: Are you testing things out are do we go with reduce_prod ?

File: python/unitytrainers/trainer_controller.py [hunk 13]
```diff
@@ -107,7 +107,10 @@ def _process_graph(self):
                 elif not self.trainers[brain_name].parameters["use_recurrent"]:
                     nodes += [scope + x for x in ["action", "value_estimate", "action_probs"]]
                 else:
-                    nodes += [scope + x for x in ["action", "value_estimate", "action_probs", "recurrent_out"]]
+                    node_list = ["action", "value_estimate", "action_probs", "recurrent_out"]
+                    if not self.trainers[brain_name].is_continuous:
+                        node_list.append("prev_action")
```
Comment (vincentpierre) [inline python/unitytrainers/trainer_controller.py]: Is there a scenario in which the developer will want to use prev_action but not recurrent ?

Comment (vincentpierre) [reply python/unitytrainers/trainer_controller.py]: Is prev_action an output of the network or is it an input ? If it is an input, then you should not need to add it here...

File: python/unitytrainers/models.py (hunk 12 above)
Comment (awjuliani) [reply python/unitytrainers/models.py]: We will keep it for now. It is theoretically correct, and balance ball trains very fast.

File: python/unitytrainers/trainer_controller.py (hunk 13 above)
Comment (awjuliani) [reply python/unitytrainers/trainer_controller.py]: Possibly, if they are using state stacking.

Comment (awjuliani) [reply python/unitytrainers/trainer_controller.py]: Gotcha. Will fix.
